source/common/ext_authz/ext_authz_impl.cc

```diff
@@ -0,0 +1,250 @@
+#include "common/ext_authz/ext_authz_impl.h"
+
+#include <chrono>
+#include <cstdint>
+#include <string>
+#include <vector>
+
+#include "envoy/access_log/access_log.h"
+#include "envoy/ssl/connection.h"
+
+#include "common/common/assert.h"
+#include "common/grpc/async_client_impl.h"
+#include "common/http/headers.h"
+
+#include "fmt/format.h"
+
+namespace Envoy {
+namespace ExtAuthz {
+
+GrpcClientImpl::GrpcClientImpl(ExtAuthzAsyncClientPtr&& async_client,
+                               const Optional<std::chrono::milliseconds>& timeout)
+    : service_method_(*Protobuf::DescriptorPool::generated_pool()->FindMethodByName(
+          "envoy.api.v2.auth.Authorization.Check")),
+      async_client_(std::move(async_client)), timeout_(timeout) {}
+
+GrpcClientImpl::~GrpcClientImpl() { ASSERT(!callbacks_); }
+
+void GrpcClientImpl::cancel() {
+  ASSERT(callbacks_ != nullptr);
+  request_->cancel();
+  callbacks_ = nullptr;
+}
+
+void GrpcClientImpl::check(RequestCallbacks& callbacks, const envoy::api::v2::auth::CheckRequest& request,
+                           Tracing::Span& parent_span) {
+  ASSERT(callbacks_ == nullptr);
+  callbacks_ = &callbacks;
+
+  request_ = async_client_->send(service_method_, request, *this, parent_span, timeout_);
+}
+
+void GrpcClientImpl::onSuccess(std::unique_ptr<envoy::api::v2::auth::CheckResponse>&& response,
+                               Tracing::Span& span) {
+  CheckStatus status = CheckStatus::OK;
+  ASSERT(response->status().code() != Grpc::Status::GrpcStatus::Unknown);
+  if (response->status().code() != Grpc::Status::GrpcStatus::Ok) {
+    status = CheckStatus::Denied;
+    span.setTag(Constants::get().TraceStatus, Constants::get().TraceUnauthz);
+  } else {
+    span.setTag(Constants::get().TraceStatus, Constants::get().TraceOk);
+  }
+
+  callbacks_->complete(status);
+  callbacks_ = nullptr;
+}
+
+void GrpcClientImpl::onFailure(Grpc::Status::GrpcStatus status, const std::string&,
+                               Tracing::Span&) {
+  ASSERT(status != Grpc::Status::GrpcStatus::Ok);
+  UNREFERENCED_PARAMETER(status);
+  callbacks_->complete(CheckStatus::Error);
+  callbacks_ = nullptr;
+}
+
+GrpcFactoryImpl::GrpcFactoryImpl(const std::string& cluster_name,
+                                 Upstream::ClusterManager& cm)
+    : cluster_name_(cluster_name), cm_(cm) {
+  if (!cm_.get(cluster_name_)) {
+    throw EnvoyException(fmt::format("unknown external authorization service cluster '{}'", cluster_name_));
+  }
+}
+
+ClientPtr GrpcFactoryImpl::create(const Optional<std::chrono::milliseconds>& timeout) {
+  return ClientPtr{new GrpcClientImpl(
+      ExtAuthzAsyncClientPtr{
+          new Grpc::AsyncClientImpl<envoy::api::v2::auth::CheckRequest,
+                                    envoy::api::v2::auth::CheckResponse>(cm_, cluster_name_)},
+      timeout)};
+}
+
+::envoy::api::v2::Address* CheckRequestGen::get_pbuf_address(const Network::Address::InstanceConstSharedPtr& instance) {
```

Comment:
Please use Envoy style, `getPufAddress`. Also, what is a `pbuf`?

Reply:
pbuf: 'protocol buffer'. Clearly I need to use a more readable name :-) Will do.

Reply:
Yeah, I think Envoyic would be `getAddress` or `getAddressFromProtobuf`.